Doc/reference/datamodel.rst

```diff
@@ -2616,6 +2616,23 @@ are awaitable.
    :term:`awaitable` objects.  For instance, :class:`asyncio.Future` implements
    this method to be compatible with the :keyword:`await` expression.
 
+   .. note::
+
+      The language doesn't place any restrictions on the type or value of the
+      objects yielded by the iterator returned by ``__await__``, as this is
+      specific to the implementation of the asynchronous execution framework
+      (e.g. :mod:`asyncio`) that will be managing the :term:`awaitable` object.
+
+      In the case of :mod:`asyncio`, user code should generally be using
+      :term:`coroutines <coroutine>`, instances of :class:`asyncio.Task`,
+      :class:`asyncio.Future`, and other :mod:`asyncio` objects to implement
```

Comment:
> user code should generally be using

> objects to implement `__await__

IMO the user code should never use `__await__`.

Reply:
In retrospect, it might make more sense to replace it with "library code" (specifically libraries using asyncio as a framework).

Reply:
This is also slightly off as asyncio should call `return yield from type(aw).__await__(aw)` or `return await aw` but actually calls `return yield from aw.__await__()`

There's also exotic objects that are Awaitables but don't have an `__await__` method. I think via c extensions

Reply:
dask/distributed uses:

```python
def __await__(self):
    async def _():
        ...

    return _().__await__()
```

To implement Awaitable classes, it's not clear to me if that's an intended use case.

Reply:
> dask/distributed uses:
> 
> ```python
> def __await__(self):
>     async def _():
>         ...
> 
>     return _().__await__()
> ```
> 
> To implement Awaitable classes, it's not clear to me if that's an intended use case.

I think it is fine. It would seem reasonable that a given function's implementation can either return the desired result, or call another function that returns the desired result. And if async function are specified to have an `__await__()` method, that's what we are doing here.

Just like when you're implementing an iterator, you can either have a method `__iter__()` that creates and returns a custom object with a `__next__()` method, or you can have `__iter__()` be a generator function. Or you can construct some other thing and call _its_ `__iter__()` method.

Reply:
> This is also slightly inaccurate as asyncio should call `return yield from type(aw).__await__(aw)` or `return await aw` but actually calls `return yield from aw.__await__()`

I can't disagree, but is this fine point relevant to the docs at hand? We're trying to specify `__await__` *without* referring to asyncio.

> There's also exotic objects that are Awaitables but don't have an `__await__` method. I think via c extensions

Can you point to an example? I'm guessing you're referring to extension types that have an `tp_as_async->am_await` slot but don't have a Python-visible `__await__` callable?